### ISSUE TITLE
DB Engine error

### ISSUE THREAD

**Issue by nyanmark** [enku/gentoo-build-publisher#4]
I followed the install guide for GBP on a clean gentoo machine, when I get to the step where I have to create a api key, I get an error when I run `gbp apikey create gbp`

```
Traceback (most recent call last):
  File "/usr/local/bin/gbp", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/gbpcli/__init__.py", line 322, in main
    return cast(int, args.func(args, GBP(args.url, auth=user_config.auth), console))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/gentoo-build-publisher/src/gentoo_build_publisher/cli/apikey.py", line 27, in handler
    return create_action(args, console)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/gentoo-build-publisher/src/gentoo_build_publisher/cli/apikey.py", line 45, in create_action
    if name in [key.name for key in publisher.repo.api_keys.list()]:
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/gentoo-build-publisher/src/gentoo_build_publisher/records/django_orm.py", line 225, in list
    return [
           ^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/query.py", line 400, in __iter__
    self._fetch_all()
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/query.py", line 1928, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/query.py", line 91, in __iter__
    results = compiler.execute_sql(
              ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 1549, in execute_sql
    sql, params = self.as_sql()
                  ^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 736, in as_sql
    extra_select, order_by, group_by = self.pre_sql_setup(
                                       ^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 84, in pre_sql_setup
    self.setup_query(with_col_aliases=with_col_aliases)
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 73, in setup_query
    self.select, self.klass_info, self.annotation_col_map = self.get_select(
                                                            ^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 296, in get_select
    sql, params = self.compile(col)
                  ^^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 546, in compile
    sql, params = node.as_sql(self, self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/expressions.py", line 1180, in as_sql
    sql = ".".join(map(compiler.quote_name_unless_alias, identifiers))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/models/sql/compiler.py", line 537, in quote_name_unless_alias
    r = self.connection.ops.quote_name(name)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gbp/lib/python3.12/site-packages/django/db/backends/dummy/base.py", line 20, in complain
    raise ImproperlyConfigured(
django.core.exceptions.ImproperlyConfigured: settings.DATABASES is improperly configured. Please supply the ENGINE value. Check settings documentation for more details.
```

**Comment by enku** [comment]
Hello @nyanmark and thanks for trying Gentoo Build Publisher. 

There should be a `settings.py` file in `/home/gbp/djangoproject/`.

Can you verify that the file exists and contains an entry like:

```python
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql',
        'NAME': 'gbp',
        'HOST': 'localhost',
        'USER': 'gbp',
        'CONN_MAX_AGE': None,
        'CONN_HEALTH_CHECKS': True,
    }
}
```

**Comment by nyanmark** [comment]
The file exists and has the values set, after the first seeing the error I also tried changing the values but it seems to ignore it.

**Comment by enku** [comment]
I will need to try this myself. I try to go through the install guide myself every time I make a significant change. The API keys was the last significant change but I may have overlooked something.

Based on https://github.com/enku/gentoo-build-publisher/issues/4#issuecomment-2224720148 I'm thinking it's maybe not even looking at the settings file. Could you try this:

```sh
DJANGO_SETTINGS_MODULE=djangoproject.settings gbp apikey create gbp
```
From the `/home/gbp` directory.

Oh... you know what. I *think* Python 3.12 changed the import mechanism so that the current directory is not automatically added to the Python path. And the last time I did an install was probably 3.11.  So if the above doesn't work try this:

```sh
PYTHONPATH=/home/gbp DJANGO_SETTINGS_MODULE=djangoproject.settings gbp apikey create gbp
```

**Comment by nyanmark** [comment]
The first variant got the error

```
ModuleNotFoundError: No module named 'djangoproject'                                                                                                                                                                                              
```

The second variant got a far more interesting error

```
  File "/home/gbp/djangoproject/settings.py", line 4, in <module>
    SECRET_KEY = os.environ['BUILD_PUBLISHER_ROOT_KEY']
                 ~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen os>", line 685, in __getitem__
KeyError: 'BUILD_PUBLISHER_ROOT_KEY'
```

when I provide the root key in the command it works

`BUILD_PUBLISHER_ROOT_KEY=... PYTHONPATH=/home/gbp DJANGO_SETTINGS_MODULE=djangoproject.settings gbp apikey create gbp`

**Comment by enku** [comment]
Thank you for reporting these issues and helping make GBP better.

1. I think understand the issue (circular dependencies) and will work on a fix this weekend.
2. That's a warning that `/etc/gentoo-build-publisher.conf` has permissions that make it readable by others. I see the reference file is world readable. So that's probably another issue on my end. You can either ignore the error or `chmod 0600 /etc/gentoo-build-publisher.conf` and it should go away.
3. What were you running when you got this error? It appears that the machine job was requiring the gentoo repo artifact. So that job needs to run (and complete) first. Perhaps it's a documentation issue that that jobs needs to complete before the machine job starts.

**Comment by nyanmark** [comment]
The permissions on the file were already 0600, that's why I was confused

```
  File: /etc/gentoo-build-publisher.conf
  Size: 361             Blocks: 8          IO Block: 4096   regular file
Device: 254,1   Inode: 5851125     Links: 1
Access: (0600/-rw-------)  Uid: ( 1000/     gbp)   Gid: ( 1000/     gbp)
```

To get the error I was following the install guide and running this step

![image](https://github.com/user-attachments/assets/d28ee4a0-b3a6-4a8f-8daa-fbaac8e253d8)

GBP created a job in jenkins and first two times it failed on the sync repos - repos/gentoo step but now it's doing the 3rd attempt and it's running just fine, updating world atm.

**Comment by enku** [comment]
Yeah, I think it probably need more emphasis in the install guide that at least one repo job needs to be complete before the machine job can run. Regarding the failures.. I'm curious as to why they failed. You can go to the job run and click "Console output" to see the raw content of the job. 
![image](https://github.com/user-attachments/assets/dd455352-89e0-4e49-bab3-0b97f15c3cd7)

Regarding the file permission warning. That's interesting. I'll take a look at that too.

**Comment by nyanmark** [comment]
The console doesn't give much info why the first two builds failed

```
Checking out Revision 69dbecd5528bed942b59b74cb8b6cb3a5d719be5 (refs/remotes/origin/master)
 > git config core.sparsecheckout # timeout=10
 > git checkout -f 69dbecd5528bed942b59b74cb8b6cb3a5d719be5 # timeout=10
Commit message: "Makefile.container: fix typo"
 > git rev-list --no-walk 69dbecd5528bed942b59b74cb8b6cb3a5d719be5 # timeout=10
[Pipeline] }
[Pipeline] // stage
[Pipeline] withEnv
[Pipeline] {
[Pipeline] withEnv
[Pipeline] {
[Pipeline] stage
[Pipeline] { (sync repos)
[Pipeline] script
[Pipeline] {
[Pipeline] readFile
[Pipeline] copyArtifacts
[Pipeline] }
[Pipeline] // script
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (build chroot)
Stage "build chroot" skipped due to earlier failure(s)
```

The latter builds, are now failing on the final step where they have to make a request to gbp.

```
+ make push
buildah unshare --mount CHROOT=base-root sh -c 'touch -r ${CHROOT}/var/cache/binpkgs/Packages packages'
time="2024-07-13T09:05:50Z" level=warning msg="The cgroupv2 manager is set to systemd but there is no systemd user session available"
time="2024-07-13T09:05:50Z" level=warning msg="For using systemd, you may need to log in using a user session"
time="2024-07-13T09:05:50Z" level=warning msg="Alternatively, you can enable lingering with: `loginctl enable-linger 473` (possibly as root)"
time="2024-07-13T09:05:50Z" level=warning msg="Falling back to --cgroup-manager=cgroupfs"
time="2024-07-13T09:05:50Z" level=warning msg="The cgroupv2 manager is set to systemd but there is no systemd user session available"
time="2024-07-13T09:05:50Z" level=warning msg="For using systemd, you may need to log in using a user session"
time="2024-07-13T09:05:50Z" level=warning msg="Alternatively, you can enable lingering with: `loginctl enable-linger 473` (possibly as root)"
time="2024-07-13T09:05:50Z" level=warning msg="Falling back to --cgroup-manager=cgroupfs"
make machine=base build=13 build.tar.gz
make[1]: Entering directory '/var/lib/jenkins/home/workspace/Gentoo/base'
make[1]: 'build.tar.gz' is up to date.
make[1]: Leaving directory '/var/lib/jenkins/home/workspace/Gentoo/base'
gbp --url=http://localhost/ pull base 13
[{'message': 'Unauthorized to resolve pull', 'locations': [{'line': 2, 'column':
3}], 'path': ['pull']}]
make: *** [Makefile:110: push] Error 1
Error when executing success post condition:
Also:   org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: a053892f-da2b-4eb4-8a66-35ff4296c3ad
hudson.AbortException: script returned exit code 2
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution.handleExit(DurableTaskStep.java:668)
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution.check(DurableTaskStep.java:614)
	at org.jenkinsci.plugins.workflow.steps.durable_task.DurableTaskStep$Execution.run(DurableTaskStep.java:555)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
```

**Comment by enku** [comment]
Actually this is a lot of info. Thanks.

All these issues have to do with the latest feature added to GBP that most people probably won't use and isn't even really needed for the purpose that I added` it.  GBP 2.1.0 (released last month) added authentication (API Keys) for web API requests.

You see GBP is mostly a web service and there is the command-line interface, `gbpcli` which is *mostly* a client to the web service. For most of GBP's life there was no authentication between client and server.  When you are running everything on one machine at home that's not exposed to the Internet then that's probably fine. And that's probably how **most** people use GBP (I say "most" but in fact, to my knowledge, you are customer number 1).

However if one wanted to expose GBP to the internet you would want some kind of authentication. That's why 2.1.0 added API keys. Certain API calls ("mutations", basically anything that makes changes) now require an API key. The reason why I added this feature, other than general best practice, is to support running remote Jenkins worker nodes (e.g. on cloud services). Now Jenkins workers already can connect to the master node via secure channels (typically ssh), however GBP builds have a post-build step where it actually makes a request to the GBP server to tell it "come pick up this new build I made".  That's the `gbp --url=http://localhost/ pull base 13` in your paste above.  Before 2.1.0 that didn't require auth. But now it does. 

You created an auth key for yourself, to use on the command line. However the Jenkins worker node also needs to have an API key because it's making a request to GBP.  You didn't know this, because I forgot to document it in the install guide :cry: . The error `'Unauthorized to resolve pull'` is coming back from GBP when Jenkins is running the `gbp pull` command.

But here's the kicker: I later learned that the post-build step that I described above doesn't actually run on the worker but runs on the master. So in actuality if you have a Jenkins worker running in the cloud as long as the master is running local to your GBP instance you don't need to expose it to the Internet. So in actuality API keys really aren't needed for the purpose I added them, though it is a good practice as I said.

For those reasons and because security adds a lot of complexity (to something that is already pretty complex) I'm going to make API keys optional. And the Install Guide will be modified to do an install with API keys disabled.

By the way congratulations on your first successful build. And the first GBP build I know of other than mine. Your `base` machine did build. It's just the post step, `gbp pull` that failed on Jenkins. However since you have created an API from your regular user, you can issue the same command yourself:

```
gbp pull base 13
```
Again, apologies for the rough edges but you are literally the first user of GBP other than me and unfortunately it came at I time where a very complex change was made.

**Comment by enku** [comment]
> No apologies needed, this is a really cool project and I am glad to be able to try it out. It was thanks to a [forum post](https://forums.gentoo.org/viewtopic-p-8701622.html?sid=4d324ed2370123317933439ed1614464) I found this. 

Oh, that was actually me posting a while back. I'd almost forgot.

I pushed some changes that enable API keys to be disabled. They're now disabled by default if you go by the current install guide, but since you've already installed you'll need to make a few changes:

### Get the latest from `master`: 
```sh
cd /home/gbp/gentoo-build-publisher
sudo -u gbp -H git pull
```

### Disable API keys in the config
```sh
echo BUILD_PUBLISHER_API_KEY_ENABLE=no >> /etc/gentoo-build-publisher.conf
```

### Restart the web service
```sh
systemctl restart gentoo-build-publisher-wsgi.service
```

After that API calls should no longer require an key and Jenkins should be able to push artifacts to GBP. See if that works.

**Comment by enku** [comment]
> I wonder if its possible to use this to automate kernel builds, I guess I can make my own kernel ebuild, have gbp build it, install that binary and make initramfs using genkernel.

You should if you use the dist kernels. But actually the reference machines should probably include a dist kernel in the `world` file. They were created before Gentoo had dist kernels my gbp instance is running in a container so it hadn't occurred to me.

The idea behind the "machines" repo is that you could (should) fork it and make it your own. So feel free to fork/branch and make modifications or add your own machines. And if you feel some change should go upstream go ahead and send me a PR.

**Comment by nyanmark** [comment]
![image](https://github.com/user-attachments/assets/d29dc620-b601-49fc-a4d3-a56b1a1259a3)

It's green now :smile: 

Thank you so much for the help

**Comment by nyanmark** [comment]
Hi, sorry for bothering you again, I finally had a chance to start figuring out the gbp machines repo, and some things a bit confusing to me.
- I am guessing every machine name has to be unique
- How would I delete a machine or repo I don't want from gbp
- How can I tell gbp which base to use other than hard coding in makefile
- What does gbp tag do

**Comment by enku** [comment]
@nyanmark I'm going to go ahead and close out this issue, but feel free to create another one should you have a concern.

> I am guessing every machine name has to be unique

I'm kind of wondering why you asked this question. In Jenkins the build names (which is a "machine" name in GBP are unique. Also since the `Jenkinsfile` uses the directory names for the config directory names must be unuique.  However "machines" in the GBP sense don't have to be 1-to-1. Think of them more as machine "profiles". For example if you had a GBP machine called "postgres"  and you have a cluster of postgres servers then they can all use the postgres machine in GBP.

> How would I delete a machine or repo I don't want from gbp

I kind of haven't made this easy intentionally. Do you mean on Jenkins or on GBP? For Jenkins you can delete them in the Jenkins UI.  For GBP there isn't a command-line way to do this yet. You can create an issue to request this feature. For now you'd have to go into the Django shell and manually delete all the builds for a given machine.

> How can I tell gbp which base to use other than hard coding in makefile

Do you mean when creating a new machine definition based on an existing one (in your gbp-machines repo)?  For example if I wanted to create a new machine called `awesome` that was based on the machine called `lighthouse`:

```
make awesome.machine base=lighthouse
```
Then edit files in the `awesome` directory to your liking.

> What does gbp tag do

Tags are a way of having more than one "published" build for a given machine. Let's take the postgres cluster example again. Say you have 100 postgres servers but you want to keep most of them "locked" on a particular build.  You can tag that build:

```
gbp tag postgres 219 stable
```
This will give the postgres.219 build a "stable" tag.  Then for your "stable" postgres nodes instead of Using "postgres" for binpkgs and repos you can use postgres@stable For example:

```
wget http://localhost/machines/postgres@stable/repos.conf
wget http://localhost/machines/gbpbox@stable/binrepos.conf
```

Now the "published" build might be `235` but for machines that are pointing at the @stable tag they will sync to `219`.  So you can have most of your machines on tried and tested `@stable` but have other machines following the latest "published" build (235).  You can also point them to a specific build, e.g. `postgres.219`, but don't do that! 

After a while you may determine that 235 is stable enough and you'll want your "stable" postgres instances to upgrade to that build. You can do this simply by moving the "stable" tag to 235.

```
gbp tag postgres 235 stable
```

Now after your "stable" instances sync they'll pick up the 235 repos and binpkgs. A given build can have more than one tag. And a machine can have many tags although the names must be unique (per machine).

You can remove a tag with the `-r` flag. You only need to provide the tag, not the number:

```
gbp tag -r postgres stable
```

Fun fact: a "published" build is just a special case of a tagged build. A published build is a build with the tag "" (empty string).  As such there is an undocumented way to "unpublish" a machine's builds:

```
gbp tag -r postgres ""
```

I hadn't realized the `gbp-machines` repos were still on the 17.1 profiles. I'll need to fix that. Thanks.
